gce_image_import_export_tests/test_suites/image_import_export_test_suite.go

```diff
@@ -0,0 +1,238 @@
+//  Copyright 2019 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+// Package importexporttestsuites contains e2e tests for image import/export cli tools
+package importexporttestsuites
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/utils/path"
+	"github.com/GoogleCloudPlatform/compute-image-tools/gce_image_import_export_tests/compute"
+	"github.com/GoogleCloudPlatform/compute-image-tools/gce_image_import_export_tests/storage"
+	"github.com/GoogleCloudPlatform/compute-image-tools/test_common/junitxml"
+	"github.com/GoogleCloudPlatform/compute-image-tools/test_common/test_config"
+)
+
+const (
+	testSuiteName = "ImageImportExportTests"
+)
+
+// TestSuite is image import test suite.
+func TestSuite(
+	ctx context.Context, tswg *sync.WaitGroup, testSuites chan *junitxml.TestSuite,
+	logger *log.Logger, testSuiteRegex, testCaseRegex *regexp.Regexp,
+	testProjectConfig *testconfig.Project) {
+
+	defer tswg.Done()
+
+	if testSuiteRegex != nil && !testSuiteRegex.MatchString(testSuiteName) {
+		return
+	}
+
+	testSuite := junitxml.NewTestSuite(testSuiteName)
+	defer testSuite.Finish(testSuites)
+	logger.Printf("Running TestSuite %q", testSuite.Name)
+	tests := runTestCases(ctx, logger, testCaseRegex, testProjectConfig)
+
+	for ret := range tests {
+		testSuite.TestCase = append(testSuite.TestCase, ret)
+	}
+
+	logger.Printf("Finished TestSuite %q", testSuite.Name)
+}
+
+func runTestCases(
+	ctx context.Context, logger *log.Logger, regex *regexp.Regexp,
+	testProjectConfig *testconfig.Project) chan *junitxml.TestCase {
+
+	imageImportDataDiskTestCase := junitxml.NewTestCase(
+		testSuiteName, fmt.Sprintf("[ImageImport] %v", "Import data disk"))
+	imageImportOSTestCase := junitxml.NewTestCase(
+		testSuiteName, fmt.Sprintf("[ImageImport] %v", "Import OS"))
+	imageImportOSFromImageTestCase := junitxml.NewTestCase(
+		testSuiteName, fmt.Sprintf("[ImageImport] %v", "Import OS from image"))
+	imageExportRawTestCase := junitxml.NewTestCase(
+		testSuiteName, fmt.Sprintf("[ImageExport] %v", "Export Raw"))
+	imageExportVMDKTestCase := junitxml.NewTestCase(
+		testSuiteName, fmt.Sprintf("[ImageExport] %v", "Export VMDK"))
+
+	testsMap := map[*junitxml.TestCase]func(
```

Comment:
You can keep reference to test function in test case so there's no need for a separate mapping from test case to test func.

Reply:
That means I need to add a field for testacase to store test func ref. However, signature of test func is different for different project (ovf test v.s. import/export test).

Reply:
What about having interface to represent test case then have different implementations for each app?

Reply:
that works, but I don't feel it benefits a lot. Please check out the latest version. Due that this "runTestCases" is shared by import and export, while ovf doesn't share it (because below for-loop has different shape), we don't get benefits from the abstraction.